File: GFG_Smallest_distinct_window.py

class Solution:
    def findSubString(self, s: str) -> int:
        n = len(s)
        distinct_chars = set(s)
        required = len(distinct_chars)

        freq = {}
        start = 0
        min_len = n
        count = 0

        for end in range(n):
            freq[s[end]] = freq.get(s[end], 0) + 1
            if freq[s[end]] == 1:
                count += 1

            while count == required:
                min_len = min(min_len, end - start + 1)
                freq[s[start]] -= 1
                if freq[s[start]] == 0:
                    count -= 1
                start += 1

        return min_len

File: GFG_Smallest_distinct_window.java

class Solution {
    public int findSubString(String s) {
        int n = s.length();
        Set<Character> distinctChars = new HashSet<>();
        
        // Count distinct characters
        for (char c : s.toCharArray()) {
            distinctChars.add(c);
        }
        int required = distinctChars.size();

        Map<Character, Integer> freq = new HashMap<>();
        int start = 0, count = 0, minLen = n;

        for (int end = 0; end < n; end++) {
            char endChar = s.charAt(end);
            freq.put(endChar, freq.getOrDefault(endChar, 0) + 1);

            if (freq.get(endChar) == 1) {
                count++;
            }

            // Try to minimize the window
            while (count == required) {
                minLen = Math.min(minLen, end - start + 1);

                char startChar = s.charAt(start);
                freq.put(startChar, freq.get(startChar) - 1);
                if (freq.get(startChar) == 0) {
                    count--;
                }
                start++;
            }
        }

        return minLen;
    }
}
